### ISSUE TITLE
[BUG] - Code executing out of order

### ISSUE THREAD

**Issue by JapanYoshi** [juanitogan/godot-firebase-lite#2]
**Describe the bug**
Even though I think I've made the code synchronous, the code is executing out of order.

**To Reproduce**
Example code and steps to reproduce the behavior:
```gdscript
func _ready():
	_init_firebase()
	_update_room_list() # SHOULD run first
	room_code = R.generate_room_code()
	var room_free = is_room_free(room_code) # SHOULD run second
	if !room_free:
		return true
	var room_made = establish_room(room_code) # SHOULD run third
	if !room_made:
		return true
	print(room_code + "Room established.")
```
Log outputs in this order:

* Firebase initialized.
* Room code is free.
* Room established.
* Room list fetched. [DRVS, HRUD, JSEL, OTUY, SPZE]

**Expected behavior**
Log should output in this order:

* Firebase initialized.
* Room list fetched.
* Room code is free.
* Room established.

**Screenshots**
If you're absolutely sure everything is installed correctly, you can skip all screenshots.  Otherwise:
 - Screenshot of your **FileSystem** tab showing your `firebase_*_lite` folders.
 - Screenshot of your **Project > Project Settings > AutoLoad** tab showing your `firebase` singleton.
 - Anything else you want to show us.

**Platform and versions (please complete the following information):**
 - Godot version: 3.5.stable.official
 - Godot Firebase Lite version: 0.1.1
 - OS: Windows 10
 - Mobile device (if applicable): N/A
 - Browser (if applicable): N/A for now, though I want to target HTML in the future

**Additional context**
I'm making a Jackbox style game where inputs are infrequent and games can be joined from across the globe via a Web browser.

**Comment by juanitogan** [comment]
It sounds like you have all the basics working, so that is good.  Excellent!

I think what you are missing is that Firebase doesn't guarantee things will be returned in the order asked, and how to cope with that in Godot.  Thus, while you may be yielding on each ref call inside your functions... if you break things out into functions like this -- and if you still want to control return order -- then you need to also yield to each function that has a yield in it (which took me more than a minute to grasp myself).  [[Well, that or become more event-driven by implementing your own signaling that controls order in its own way.]]  Otherwise, no, you are not synchronous, because your `_ready` code is exiting each function as soon as it hits a `yield()` in it, and moving onto the next `_ready` line.  The yields then finish later in whatever order they finish.

So, in other words, I'm guessing you need to yield to each of `_update_room_list`, `is_room_free`, and `establish_room` like so: `yield(_update_room_list(), "completed")` etc.

You can look at Fetch, Put, Push, etc., in firebase_database_lite/reference.gd for some simple examples of how to write functions with yields in them, that you then yield to when calling them.  Whether you simply `return yield(ref.update(...), "completed")` or `return` something custom, that is up to you.

[[Again, just to be thorough about Godot features and programming patterns, signalling is another option to this, but I find signalling overkill when you just needs a few things done in order.]]

**Comment by JapanYoshi** [comment]
I tried var room_free = yield(is_room_free(room_code), "completed") but it says "First argument of yield() not of type object".

I yield multiple times in the function, so I'm not sure what I'm supposed to do with this.

```gdscript
func is_room_free(room_code: String) -> bool:
	if not(room_code in room_codes):
		print("Room code is free.")
		return true
	# Room code is in the list.
	var result
	var ref: FirebaseReference = db.get_reference_lite(
		"rooms/" + room_code
	)
	result = yield(ref.fetch(), "completed")
	if result is FirebaseError:
		printerr("Error on checking rooms/" + room_code + ".")
		return false
	# Is this an old room?
	if Time.get_unix_time_from_system() - 5 * 60 < result.value["lastUpdate"]:
		print("Room has been updated in the last 5 minutes. Game is probably in progress.")
		return false
	print("Room has not been updated in 5 minutes. Can safely delete, but don't use just in case.")
	ref = db.get_reference_lite("roomCodes/" + room_code)
	result = yield(ref.update(false), "completed")
	if result is FirebaseError:
		printerr("Could not update roomCodes/" + room_code + " to True. Aborted.")
	return false
```

**Comment by juanitogan** [comment]
Okay, yes, I had to google that one.  `yield` requires an object as the first param; or, in this case, a function that returns an object.  Your function returns bool.  Change it to `-> Object` and return `OK` and/or `FAILED` instead, or something along those lines.  This is just one of the many funky things about GDScript I forget because I don't work with it regularly.

[[You could maybe return null instead and still use `if !room_free` but I don't recall if that works or not.  Regardless, GDScript prefers the pattern of returning `OK` and comparing against that.  I think I ran into the same thing and eventually gave in to using `OK`.  And, yes, somehow, error enums qualify as objects even though it doesn't say this anywhere.  Very weird.]]
